enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java

```diff
@@ -0,0 +1,365 @@
+    {
+        this.log = log;
+        this.availabilityGuard = availabilityGuard;
+        workQueue = new LinkedBlockingQueue<>();
+    }
+
+    <E extends Entity> BinaryLatch updatePropertyData( Map<Long,Map<String,Object>> state, WritableFulltext index )
+            throws
+            IOException
+    {
+        BinaryLatch completedLatch = new BinaryLatch();
+        FulltextIndexUpdate update = () ->
+        {
+            PartitionedIndexWriter indexWriter = index.getIndexWriter();
+            for ( Map.Entry<Long,Map<String,Object>> stateEntry : state.entrySet() )
+            {
+                Set<String> indexedProperties = index.properties();
+                if ( !Collections.disjoint( indexedProperties, stateEntry.getValue().keySet() ) )
+                {
+                    long entityId = stateEntry.getKey();
+                    Stream<Map.Entry<String,Object>> entryStream = stateEntry.getValue().entrySet().stream();
+                    Predicate<Map.Entry<String,Object>> relevantForIndex =
+                            entry -> indexedProperties.contains( entry.getKey() );
+                    Map<String,Object> allProperties = entryStream.filter( relevantForIndex ).collect(
+                            Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) );
+
+                    if ( !allProperties.isEmpty() )
+                    {
+                        updateDocument( indexWriter, entityId, allProperties );
+                    }
+                }
+            }
+            return Pair.of( index, completedLatch );
+        };
+
+        enqueueUpdate( update );
+        return completedLatch;
+    }
+
+    private static void updateDocument( PartitionedIndexWriter indexWriter, long entityId,
+                                        Map<String,Object> properties ) throws IOException
+    {
+        Document document = documentRepresentingProperties( entityId, properties );
+        indexWriter.updateDocument( newTermForChangeOrRemove( entityId ), document );
+    }
+
+    <E extends Entity> BinaryLatch removePropertyData( Iterable<PropertyEntry<E>> propertyEntries,
+                                                       Map<Long,Map<String,Object>> state,
+                                                       WritableFulltext index ) throws IOException
+    {
+        BinaryLatch completedLatch = new BinaryLatch();
+        FulltextIndexUpdate update = () ->
+        {
+            for ( PropertyEntry<E> propertyEntry : propertyEntries )
+            {
+                if ( index.properties().contains( propertyEntry.key() ) )
+                {
+                    long entityId = propertyEntry.entity().getId();
+                    Map<String,Object> allProperties = state.get( entityId );
+                    if ( allProperties == null || allProperties.isEmpty() )
+                    {
+                        index.getIndexWriter().deleteDocuments( newTermForChangeOrRemove( entityId ) );
```

Comment:
this thing still creates index writer each time

Reply:
The writer is cached in the `index` now.